File: src/eQuantic.UI.Runtime/src/shared/components/CodeDiffSource.ts
import { CodeDiffer, CodeDiffGap, CodeDocument, CodeLineChange, CodePatchFile } from "../runtime-exports";

export class CodeDiffSource {
    constructor(original: CodeDocument, originalLineCount: number, modified: CodeDocument, modifiedLineCount: number, changes: CodeLineChange[], originalNumbers: number[] | null, modifiedNumbers: number[] | null, gaps: CodeDiffGap[], props?: any) {
        this._originalNumbers = null;
        this._modifiedNumbers = null;
        this.original = original;
        this.originalLineCount = originalLineCount;
        this.modified = modified;
        this.modifiedLineCount = modifiedLineCount;
        this.changes = changes;
        this._originalNumbers = originalNumbers;
        this._modifiedNumbers = modifiedNumbers;
        this.gaps = gaps;
        if (props && typeof props === 'object') Object.assign(this, props);
    }

    _originalNumbers: number[] | null;
    _modifiedNumbers: number[] | null;
    declare original: CodeDocument;
    declare modified: CodeDocument;
    originalLineCount: number = 0;
    modifiedLineCount: number = 0;
    declare changes: CodeLineChange[];
    declare gaps: CodeDiffGap[];

    originalNumber(line: number) {
        let numbers: any; return (numbers = this._originalNumbers) != null ? numbers[line] : line + 1;
    }

    modifiedNumber(line: number) {
        let numbers: any; return (numbers = this._modifiedNumbers) != null ? numbers[line] : line + 1;
    }

    static fromTexts(original: string, modified: string) {
        return CodeDiffSource.fromDocuments(CodeDocument.fromText(original), CodeDocument.fromText(modified));
    }

    static fromDocuments(original: CodeDocument, modified: CodeDocument) {
        return new CodeDiffSource(original, original.lineCount, modified, modified.lineCount, CodeDiffer.compare(original, modified), null, null, []);
    }

    originalLineOf(modifiedLine: number) {
        let low = 0;
        let high = this.changes.length - 1;
        let found = -1;
        while (low <= high) {
            let middle = Math.trunc((low + high) / 2);
            if (this.changes[middle].modifiedStart <= modifiedLine) {
                found = middle;
                low = middle + 1;
            } else high = middle - 1;
        }
        if (found < 0) return modifiedLine;
        let change = this.changes[found];
        if (modifiedLine < change.modifiedStart + change.modifiedCount) return -1;
        return change.originalStart + change.originalCount + (modifiedLine - change.modifiedStart - change.modifiedCount);
    }

    static fromPatch(file: CodePatchFile) {
        let originalLines: string[] = [];
        let modifiedLines: string[] = [];
        let originalNumbers: number[] = [];
        let modifiedNumbers: number[] = [];
        let changes: CodeLineChange[] = [];
        let gaps: CodeDiffGap[] = [];
        let originalEnd = 0;
        let modifiedEnd = 0;
        for (const hunk of file.hunks) {
            let skippedOriginal = hunk.originalStart - originalEnd;
            let skippedModified = hunk.modifiedStart - modifiedEnd;
            if (skippedOriginal > 0 || skippedModified > 0) gaps.push(new CodeDiffGap(originalLines.length, modifiedLines.length, skippedOriginal, skippedModified, hunk.header));
            let originalNumber = hunk.originalStart;
            let modifiedNumber = hunk.modifiedStart;
            let i = 0;
            while (i < hunk.lines.length) {
                if (hunk.lines[i].kind === 'context') {
                    originalLines.push(hunk.lines[i].text);
                    originalNumbers.push(++originalNumber);
                    modifiedLines.push(hunk.lines[i].text);
                    modifiedNumbers.push(++modifiedNumber);
                    i++;
                    continue;
                }
                let originalStart = originalLines.length;
                let modifiedStart = modifiedLines.length;
                while (i < hunk.lines.length && hunk.lines[i].kind !== 'context') {
                    if (hunk.lines[i].kind === 'removed') {
                        originalLines.push(hunk.lines[i].text);
                        originalNumbers.push(++originalNumber);
                    } else {
                        modifiedLines.push(hunk.lines[i].text);
                        modifiedNumbers.push(++modifiedNumber);
                    }
                    i++;
                }
                let originalCount = originalLines.length - originalStart;
                let modifiedCount = modifiedLines.length - modifiedStart;
                changes.push(new CodeLineChange(originalStart, originalCount, modifiedStart, modifiedCount, CodeDiffer.innerChanges(originalLines, originalStart, originalCount, modifiedLines, modifiedStart, modifiedCount)));
            }
            originalEnd = hunk.originalStart + hunk.originalCount;
            modifiedEnd = hunk.modifiedStart + hunk.modifiedCount;
        }
        return new CodeDiffSource(CodeDocument.fromLines(originalLines), originalLines.length, CodeDocument.fromLines(modifiedLines), modifiedLines.length, changes, originalNumbers, modifiedNumbers, gaps);
    }
}

File: src/eQuantic.UI.Code/Diff/CodeDiffSource.cs
namespace eQuantic.UI.Code;

/// <summary>
/// What a diff view shows: an <see cref="Original"/> and a <see cref="Modified"/> text, what differs
/// between them (<see cref="Changes"/>), and the number each line has in its file.
/// <para>
/// From two texts (<see cref="FromTexts"/>) they are whole files, a line's number is its index plus
/// one, and the changes are <see cref="CodeDiffer"/>'s. From a patch (<see cref="FromPatch"/>) they
/// are the lines the patch quotes, hunk after hunk, numbered as the patch numbers them; the changes
/// are the ones the patch marks, a run of removed and added lines between two lines of context
/// being one change, with its words compared; and between two hunks a <see cref="CodeDiffGap"/> says
/// how many lines of the file the patch left out.
/// </para>
/// </summary>
public sealed class CodeDiffSource
{
    private readonly IReadOnlyList<int>? _originalNumbers;
    private readonly IReadOnlyList<int>? _modifiedNumbers;

    private CodeDiffSource(CodeDocument original, int originalLineCount, CodeDocument modified, int modifiedLineCount,
        IReadOnlyList<CodeLineChange> changes, IReadOnlyList<int>? originalNumbers, IReadOnlyList<int>? modifiedNumbers,
        IReadOnlyList<CodeDiffGap> gaps)
    {
        Original = original;
        OriginalLineCount = originalLineCount;
        Modified = modified;
        ModifiedLineCount = modifiedLineCount;
        Changes = changes;
        _originalNumbers = originalNumbers;
        _modifiedNumbers = modifiedNumbers;
        Gaps = gaps;
    }

    /// <summary>The original text, or the lines of it a patch quotes.</summary>
    public CodeDocument Original { get; }

    /// <summary>The modified text, or the lines of it a patch quotes.</summary>
    public CodeDocument Modified { get; }

    /// <summary>
    /// How many lines each side has for the view, which is the document's own count for a whole text
    /// and 0 for the side of a patch that quotes nothing, a file added or deleted, where the document
    /// still holds its one empty line. The rows of a view are counted on these.
    /// </summary>
    public int OriginalLineCount { get; }

    /// <inheritdoc cref="OriginalLineCount"/>
    public int ModifiedLineCount { get; }

    /// <summary>What differs, in the lines of <see cref="Original"/> and <see cref="Modified"/>.</summary>
    public IReadOnlyList<CodeLineChange> Changes { get; }

    /// <summary>The runs of the file a patch left out: none for two whole texts.</summary>
    public IReadOnlyList<CodeDiffGap> Gaps { get; }

    /// <summary>The number line <paramref name="line"/> of <see cref="Original"/> has in its file,
    /// counted from 1 as a gutter counts.</summary>
    public int OriginalNumber(int line) => _originalNumbers is { } numbers ? numbers[line] : line + 1;

    /// <summary>The number line <paramref name="line"/> of <see cref="Modified"/> has in its file.</summary>
    public int ModifiedNumber(int line) => _modifiedNumbers is { } numbers ? numbers[line] : line + 1;

    /// <summary>Two whole texts, compared.</summary>
    public static CodeDiffSource FromTexts(string original, string modified)
    {
        var before = CodeDocument.FromText(original);
        var after = CodeDocument.FromText(modified);
        return new CodeDiffSource(before, before.LineCount, after, after.LineCount, CodeDiffer.Compare(before, after),
            null, null, []);
    }

    /// <summary>The lines one file of a patch quotes, and the changes it marks.</summary>
    public static CodeDiffSource FromPatch(CodePatchFile file)
    {
        var originalLines = new List<string>();
        var modifiedLines = new List<string>();
        var originalNumbers = new List<int>();
        var modifiedNumbers = new List<int>();
        var changes = new List<CodeLineChange>();
        var gaps = new List<CodeDiffGap>();
        var originalEnd = 0;
        var modifiedEnd = 0;
        foreach (var hunk in file.Hunks)
        {
            var skippedOriginal = hunk.OriginalStart - originalEnd;
            var skippedModified = hunk.ModifiedStart - modifiedEnd;
            if (skippedOriginal > 0 || skippedModified > 0)
                gaps.Add(new CodeDiffGap(originalLines.Count, modifiedLines.Count, skippedOriginal, skippedModified));

            var originalNumber = hunk.OriginalStart;
            var modifiedNumber = hunk.ModifiedStart;
            var i = 0;
            while (i < hunk.Lines.Count)
            {
                if (hunk.Lines[i].Kind == CodePatchLineKind.Context)
                {
                    originalLines.Add(hunk.Lines[i].Text);
                    originalNumbers.Add(++originalNumber);
                    modifiedLines.Add(hunk.Lines[i].Text);
                    modifiedNumbers.Add(++modifiedNumber);
                    i++;
                    continue;
                }
                // A run of removed and added lines, up to the next line of context, is one change.
                var originalStart = originalLines.Count;
                var modifiedStart = modifiedLines.Count;
                while (i < hunk.Lines.Count && hunk.Lines[i].Kind != CodePatchLineKind.Context)
                {
                    if (hunk.Lines[i].Kind == CodePatchLineKind.Removed)
                    {
                        originalLines.Add(hunk.Lines[i].Text);
                        originalNumbers.Add(++originalNumber);
                    }
                    else
                    {
                        modifiedLines.Add(hunk.Lines[i].Text);
                        modifiedNumbers.Add(++modifiedNumber);
                    }
                    i++;
                }
                var originalCount = originalLines.Count - originalStart;
                var modifiedCount = modifiedLines.Count - modifiedStart;
                changes.Add(new CodeLineChange(originalStart, originalCount, modifiedStart, modifiedCount,
                    CodeDiffer.InnerChanges(originalLines, originalStart, originalCount, modifiedLines, modifiedStart, modifiedCount)));
            }
            originalEnd = hunk.OriginalStart + hunk.OriginalCount;
            modifiedEnd = hunk.ModifiedStart + hunk.ModifiedCount;
        }
        return new CodeDiffSource(CodeDocument.FromLines(originalLines), originalLines.Count,
            CodeDocument.FromLines(modifiedLines), modifiedLines.Count, changes, originalNumbers, modifiedNumbers, gaps);
    }
}
